circus-train-s3-s3-copier/src/main/java/com/hotels/bdp/circustrain/s3s3copier/aws/JceksAmazonS3ClientFactory.java

```diff
@@ -58,6 +67,25 @@ public AmazonS3 newInstance(AmazonS3URI uri, S3S3CopierOptions s3s3CopierOptions
     }
   }
 
+  public AmazonS3 newTargetInstance(AmazonS3URI uri, S3S3CopierOptions s3s3CopierOptions) {
+    AmazonS3 globalClient;
+
+    String assumedRole = s3s3CopierOptions.getAssumedRole();
+    if (assumedRole != null) {
```

Comment:
There's a lot of code in this method which is duplicated elsewhere in this class, feels like a lot of overlap between methods. To me it feels like there should be two code paths - one where the role is assumed (the new stuff) and one where it isn't (what was there before). Can we try isolate the differences into a few new methods based on an if statement on `assumedRole` being null and everything else can stay as it was.

Reply:
I was trying out few things to refactor this and ended up creating this PR: https://github.com/HotelsDotCom/circus-train/pull/159

Basically, the switch now happens only at `getCredentialsProviderChain` function. Have a look and we can improve it further.